adapters/aduptech/aduptech.go

```diff
@@ -0,0 +1,156 @@
+	var extraInfo ExtraInfo
+	if err := jsonutil.Unmarshal([]byte(config.ExtraAdapterInfo), &extraInfo); err != nil {
+		return nil, fmt.Errorf("invalid extra info: %v", err)
+	}
+
+	bidder := &adapter{
+		endpoint:  config.Endpoint,
+		extraInfo: extraInfo,
+	}
+
+	return bidder, nil
+}
+
+func (a *adapter) MakeRequests(request *openrtb2.BidRequest, reqInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	for i := range request.Imp {
+		imp := &request.Imp[i]
+		// Check if imp comes with bid floor amount defined in a foreign currency
+		if imp.BidFloor > 0 && imp.BidFloorCur != "" && strings.ToUpper(imp.BidFloorCur) != a.extraInfo.TargetCurrency {
+
+			convertedValue, err := a.convertCurrency(imp.BidFloor, imp.BidFloorCur, reqInfo)
+			if err != nil {
+				return nil, []error{err}
+			}
+
+			imp.BidFloorCur = a.extraInfo.TargetCurrency
+			imp.BidFloor = convertedValue
+		}
+	}
+
+	requestJSON, err := jsonutil.Marshal(request)
+	if err != nil {
+		return nil, []error{err}
+	}
+
+	requestData := &adapters.RequestData{
+		Method: http.MethodPost,
+		Uri:    a.endpoint,
+		Body:   requestJSON,
+		ImpIDs: openrtb_ext.GetImpIDs(request.Imp),
+	}
+
+	return []*adapters.RequestData{requestData}, nil
+}
+
+func (a *adapter) convertCurrency(value float64, cur string, reqInfo *adapters.ExtraRequestInfo) (float64, error) {
+	convertedValue, err := reqInfo.ConvertCurrency(value, cur, a.extraInfo.TargetCurrency)
+
+	if err != nil {
+		var convErr currency.ConversionNotFoundError
+		if errors.As(err, &convErr) {
+
+			// try again by first converting to USD
+			// then convert to target_currency
+			convertedValue, err = reqInfo.ConvertCurrency(value, cur, "USD")
+
+			if err != nil {
+				return 0, err
+			}
+
+			convertedValue, err = reqInfo.ConvertCurrency(convertedValue, "USD", a.extraInfo.TargetCurrency)
+
+			if err != nil {
+				return 0, err
+			}
+		} else {
+			return 0, err
+		}
+	}
+	return convertedValue, nil
+}
+
+func (a *adapter) MakeBids(request *openrtb2.BidRequest, requestData *adapters.RequestData, responseData *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+	switch responseData.StatusCode {
+	case http.StatusNoContent:
+		return nil, nil
+	case http.StatusBadRequest:
+		return nil, []error{&errortypes.BadInput{
+			Message: fmt.Sprintf("Unexpected status code: 400. Run with request.debug = 1 for more info."),
+		}}
+	case http.StatusOK:
+		// we don't need to do anything here and this is just so we can use the default case for the unexpected status code
+		break
+	default:
+		return nil, []error{&errortypes.BadServerResponse{
+			Message: fmt.Sprintf("Unexpected status code: %d. Run with request.debug = 1 for more info.", responseData.StatusCode),
+		}}
+
+	}
+
+	var response openrtb2.BidResponse
+	if err := jsonutil.Unmarshal(responseData.Body, &response); err != nil {
+		return nil, []error{err}
```

Comment:
Please add a supplemental JSON test where your mock response returns a bid response that fails to unmarshal successfully. This can be achieved by setting one of the fields to an incorrect data type.